src/ArborX_DBSCAN.hpp

```diff
@@ -147,58 +138,164 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
+                                                                  core_points});
+      Kokkos::Profiling::popRegion();
+    }
+    else
+    {
+      // Determine core points
+      Kokkos::Timer timer_local;
+      timer_start(timer_local);
+      Kokkos::Profiling::pushRegion("ArborX::dbscan::clusters::num_neigh");
+      Kokkos::View<int *, MemorySpace> num_neigh(
+          "ArborX::dbscan::num_neighbors", n);
+      bvh.query(exec_space, predicates,
+                Details::FDBSCANNumNeighEarlyExitCallback<MemorySpace>{
+                    num_neigh, core_min_size});
+      Kokkos::Profiling::popRegion();
+      elapsed["neigh"] = timer_seconds(timer_local);
+
+      using CorePoints = DBSCAN::DBSCANCorePoints<MemorySpace>;
+
+      // Perform the queries and build clusters through callback
+      timer_start(timer_local);
+      Kokkos::Profiling::pushRegion("ArborX::dbscan::clusters:query");
+      bvh.query(exec_space, predicates,
+                Details::FDBSCANCallback<MemorySpace, CorePoints>{
+                    labels, CorePoints{num_neigh, core_min_size}});
+      Kokkos::Profiling::popRegion();
+      elapsed["query"] = timer_seconds(timer_local);
+    }
   }
-  else
+  else if (parameters._implementation ==
+           DBSCAN::Implementation::FDBSCAN_DenseBox)
   {
-    // Determine core points
-    Kokkos::Timer timer_local;
-    timer_start(timer_local);
-    Kokkos::Profiling::pushRegion("ArborX::dbscan::clusters::num_neigh");
-    Kokkos::View<int *, MemorySpace> num_neigh("ArborX::dbscan::num_neighbors",
-                                               n);
-    bvh.query(exec_space, predicates,
-              DBSCAN::NumNeighEarlyExitCallback<MemorySpace>{num_neigh,
-                                                             core_min_size});
-    Kokkos::Profiling::popRegion();
-    elapsed["neigh"] = timer_seconds(timer_local);
+    auto const predicates = buildPredicates(primitives, eps);
+
+    // Find dense boxes
+    timer_start(timer);
+    Kokkos::Profiling::pushRegion("ArborX::dbscan::dense_cells");
+    Box bounds;
+    Details::TreeConstruction::calculateBoundingBoxOfTheScene(
+        exec_space, primitives, bounds);
+    float const h = eps / std::sqrt(3); // 3D (for 2D change to std::sqrt(2))
+    auto const &min_corner = bounds.minCorner();
+    auto const &max_corner = bounds.maxCorner();
+    size_t nx = std::ceil((max_corner[0] - min_corner[0]) / h);
+    size_t ny = std::ceil((max_corner[1] - min_corner[1]) / h);
+    size_t nz = std::ceil((max_corner[2] - min_corner[2]) / h);
+    if (verbose)
+      printf("h = %e, nx = %zu, ny = %zu, nz = %zu\n", h, nx, ny, nz);
+
+    auto cell_indices = Details::computeCellIndices(exec_space, primitives, h,
+                                                    nx, ny, nz, bounds);
+    auto permute = Details::sortObjects(exec_space, cell_indices);
+
+    auto mixed_offsets = Details::computeMixedOffsets(exec_space, core_min_size,
```

Comment:
It's a bit hard to explain. Essentially, we want to construct a tree on a combination of individual points (belonging to non-dense cells) and boxes of dense cells. In this sense, the primitives are mixed. Mixed offsets is essentially offsets in the sorted cells indices array corresponding to mixed primitives. For example, if you have `[a a b b b]` cell indices, and minpts = 3, it will result in the mixed offsets `[0, 1, 2, 5]` mixed offsets.